packages/noya-state/src/selectors/gradientSelectors.ts

```diff
@@ -177,3 +236,86 @@ export function isPointerOnGradientLine(state: ApplicationState, point: Point) {
     selectedLayerGradientPoints[selectedLayerGradientPoints.length - 1].point,
   ]);
 }
+
+export function getCircleTangentSquare(
+  center: Point,
+  point: Point,
+  ellipseLength: number,
+) {
+  const height = distance(center, point) * 2;
+  const width = ellipseLength === 0 ? height : height * ellipseLength;
+
+  const theta =
+    Math.atan2(point.y - center.y, point.x - center.x) - Math.PI / 2;
+
+  return {
+    rectangle: {
+      x: center.x - width / 2,
+      y: center.y,
+      height,
+      width,
+    },
+    theta,
+  };
+}
+
+export function isPointerOnGradientEllipseEditor(
+  state: ApplicationState,
+  point: Point,
+) {
+  const selectedLayerGradientPoints = getSelectedGradientStopPoints(state);
+  if (!state.selectedGradient || !selectedLayerGradientPoints) return false;
+  const selectedGradient = getSelectedGradient(
+    getCurrentPage(state),
+    state.selectedGradient,
+  );
+  if (!selectedGradient) return false;
+
+  const center = selectedLayerGradientPoints[0].point;
+  const lastPoint =
+    selectedLayerGradientPoints[selectedLayerGradientPoints.length - 1].point;
+
+  const { rectangle, theta } = getCircleTangentSquare(
+    center,
+    lastPoint,
+    selectedGradient.elipseLength,
+  );
+
+  const position = AffineTransform.rotate(theta, center.x, center.y).applyTo({
+    x: rectangle.x,
+    y: rectangle.y,
+  });
+
+  return isPointInRange(point, position, SELECTED_GRADIENT_POINT_RADIUS);
+}
+
+export function getAngularGradientCircle(state: ApplicationState) {
+  if (!state.selectedGradient) return;
+
+  const { layerId } = state.selectedGradient;
+  const page = getCurrentPage(state);
+  const indexPath = Layers.findIndexPath(
+    page,
+    (layer) => layer.do_objectID === layerId,
+  );
+
+  if (!indexPath) return;
+
+  const parent = getParentLayer(page, indexPath);
```

Comment:
We need to consider every ancestor here, e.g.

```ts
const layer = Layers.access(page, indexPath)
const transform = getLayerTransformAtIndexPath(page, indexPath)
    .transform(getLayerFlipTransform(layer))
    .transform(getLayerRotationTransform(layer))
const bounds = createBounds(layer.frame)
const center = transform.applyTo({ x: bounds.midX, y: bounds.midY })
```

Otherwise, the same problem will happen if there's more than one parent:

![Screen Shot 2021-08-12 at 10 54 49 AM](https://user-images.githubusercontent.com/1198882/129249481-778c11c8-fd68-4aee-8adb-3c9fa946bc75.png)

We should also handle rotations. I think the way to do that with the fewest changes would be to return a `rotation` value from this function.

```ts
const line = transform.applyTo({ x: bounds.maxX, y: bounds.midY })
const rotation = Math.atan2(line.y - center.y, line.x - center.x)

return { center, radius, rotation }
```

Then we can use the rotation to position the gradient stops, e.g. in `getSelectedGradientStopPoints`:

```ts
const radians = stop.position * Math.PI * 2 + circle.rotation
```

Right now, we're not handling rotation and flip, so the stops will appear in the wrong place:

![Screen Shot 2021-08-12 at 11 41 32 AM](https://user-images.githubusercontent.com/1198882/129251113-2c6bb310-68d8-4ef3-b5c8-ee394a1ba68d.png)